README.md
# DNB Design System 👨‍🎨

![eufemia_flutter](https://github.com/dnb-asa/eufemia_flutter/workflows/Dart%20CI/badge.svg) ![pub](https://img.shields.io/pub/v/eufemia?logo=flutter) [![style: pedantic](https://img.shields.io/badge/style-pedantic-9cf)](https://github.com/dart-lang/pedantic)

![preview](https://github.com/dnb-asa/eufemia_flutter/blob/master/assets/preview.png?raw=true)

> **DISCLAIMER**: This package is in an early stage

## Eufemia

![pub](https://img.shields.io/pub/v/eufemia?logo=flutter)

### Setup

Fonts and icons needs to be bundled with your app. Add the following to your `pubspec.yaml`:

```yaml
flutter:
  uses-material-design: true

  fonts:
    - family: EufemiaIcons
      fonts:
        - asset: packages/eufemia/assets/EufemiaIcons.ttf

    - family: DNB Sans
      fonts:
        - asset: packages/eufemia/fonts/DNB-Light.otf
          weight: 300
        - asset: packages/eufemia/fonts/DNB-Regular.otf
          weight: 400
        - asset: packages/eufemia/fonts/DNB-Medium.otf
          weight: 500
        - asset: packages/eufemia/fonts/DNB-Bold.otf
          weight: 600
        - asset: packages/eufemia/fonts/DNB-LightItalic.otf
          weight: 300
          style: italic
        - asset: packages/eufemia/fonts/DNB-RegularItalic.otf
          weight: 400
          style: italic
        - asset: packages/eufemia/fonts/DNB-MediumItalic.otf
          weight: 500
          style: italic
        - asset: packages/eufemia/fonts/DNB-BoldItalic.otf
          weight: 600
          style: italic

    - family: DNB Mono
      fonts:
        - asset: packages/eufemia/fonts/DNBMono-Light.otf
          weight: 300
        - asset: packages/eufemia/fonts/DNBMono-Regular.otf
          weight: 400
        - asset: packages/eufemia/fonts/DNBMono-Medium.otf
          weight: 500
        - asset: packages/eufemia/fonts/DNBMono-Bold.otf
          weight: 600
        - asset: packages/eufemia/fonts/DNBMono-LightItalic.otf
          weight: 300
          style: italic
        - asset: packages/eufemia/fonts/DNBMono-RegularItalic.otf
          weight: 400
          style: italic
        - asset: packages/eufemia/fonts/DNBMono-MediumItalic.otf
          weight: 500
          style: italic
        - asset: packages/eufemia/fonts/DNBMono-BoldItalic.otf
          weight: 600
          style: italic
```

### How to use

Wrap your app in an `Eufemia` widget, and provide the `EufemiaData` with the `palette`, `spacing`, `button` and optionally `darkPalette` parameters:

```dart
Eufemia(
  data: EufemiaData(
    palette: EufemiaPaletteData.standard(),
    spacing: EufemiaSpacingData.fallback(),
    button: EufemiaButtonThemeData.fallback(),
    darkPalette: EufemiaPaletteData.dark(),
  ),
),
```

To use Material widgets, e.g. in a `MaterialApp` tree, use the provided adapter to provide a Material Theme:

```dart
final adapter = EufemiaThemeAdapter(context);

return MaterialApp(
  theme: adapter.theme,
  ...
);
```

The Eufemia provider with all the theming data can be accessed using

```dart
final eufemia = Eufemia.of(context)
```

## Components

There are multiple UI components

## Palette

**This API is subject to change**

The `EufemiaPaletteData` class is a collection of colors that are used throughout the library of components. You can create your own, or use several presets, like the `saga`, `standard`, `dark` and `privateBanking` factories.

The palette provider can be accessed using

```dart
final palette = EufemiaPalette.of(context)
```

## Typography

If using the included themes, fonts are automatically set up. Otherwise, specify the `fontFamily` and `package` properties:

```dart
TextStyle(
  fontFamily: 'DNB Sans',
  package: 'eufemia',
)
```

The typography provider can be accessed using

```dart
final typography = EufemiaTypography.of(context)
```

## Spacing

The `EufemiaSpacingData` class defines the different kinds of spacing used throughout the app. By default, it provides the standard spacings defined by the Eufemia design system:

- `extraSmall: 4.0`
- `small: 8.0`
- `medium : 16.0`
- `large: 24.0`
- `extraLarge: 32.0`
- `extraExtraLarge: 40.0`

The spacing provider can be accessed using

```dart
final spacing = EufemiaSpacing.of(context)
```

## Desktop

Desktop targets are supported but not currently tested extensively, but may be prioritized in the future.
a

## Web

Web targets are supported, but the Web APIs are unstable, and the target is not tested at all. If building a web application, check out [eufemia for web](https://github.com/dnbexperience/eufemia) for a web native version.


CHANGELOG.md
## 3.0.0

**Breaking changes**

- Combine libraries to single package

**Migration guide:**
Change import statements:

- `package:eufemia_components/eufemia_components.dart` → `package:eufemia/components.dart`
- `package:eufemia_palette/eufemia_palette.dart` → `package:eufemia/palette.dart`
- `package:eufemia_typography/eufemia_typography.dart` → `package:eufemia/typography.dart`
- `package:eufemia_spacing/eufemia_spacing.dart` → `package:eufemia/spacing.dart`

## 2.0.3

- Add `isStandaloneApp` property to `EufemiaData` to allow for add-to-app modules

## 2.0.2

- Add `textSelectionHandleColor` to `EufemiaAdapter`

## 2.0.1

- Bump dependencies

## 2.0.0

**Breaking changes**

- Update example
- Bump `eufemia_components` to 2.0.0:

> Rename button namespace to avoid conflict with new Material buttons:
>
> - `PrimaryButton` → `EufemiaPrimaryButton`
> - `SecondaryButton` → `EufemiaSecondaryButton`
> - `TextButton` → `EufemiaTextButton`
> - `VividButton` → `EufemiaVividButton`

## 1.0.5+1

- Update readme

## 1.0.5

- Add `fromTextStyle` factory on `Shimmer` component
- Fix sync issues with `EufemiaTabBar` component

## 1.0.4

- Add new PageIndicator
- Fix Navbar bugs
- Set EufemiaTab active parameter to default false
- Update example app
- Add EufemiaTabBar

## 1.0.3

- Update example
- Rename theme.dart to adapter.dart
- Update eufemia_palette and eufemia_components constraint

## 1.0.2

- Allow for non-text titles on list components
- Update version constraints

## 1.0.0

- Add theme providers
- Refactor components to make use of theme providers
- Split project into separate packages

## 0.6.0+1

- Readme badges

## 0.6.0

- Adds `MessageBox` component
- Adds `borderRadius` property on `ColoredIcon`

## 0.5.0

- Changes color on tapped buttons
- Finetunes padding for SegmentedControl
- Adds new GradientWidget component
- Adds new language features
- Tweaks to StaticList, Cell, ScrollableList, Shimmer, shadows, sheets

## 0.4.0

- Added Segmented Control component
- Refactored components
- Added adaptive StaticList/ScrollableList
- Tweaked dark mode colors
- Added extensions
- Added flippable parameter to PaymentCard

### **BREAKING CHANGES**

- TabItem renamed to TabItemData
- Updated PaymentCard API

## 0.3.0

- Added DNB Sans variants

### **BREAKING CHANGES**

- Replaced Menlo-Regular with DNB Mono

## 0.2.8

- Added macOS, web examples
- Fixed InputField borders in dark mode
- Updated to latest Material TextTheme API

## 0.2.7

- Added InputField
- Added Checkmark

## 0.2.6+1

- Fixed borderRadius in circular shimmer

## 0.2.6

- Added license

## 0.2.5

- Introduced circular/custom-shaped shimmer

## 0.2.4

- Switched to native SnackBar for toasts

## 0.2.3

- Added shimmer/skeleton
- Added shimmerList to Cell

## 0.2.2+2

- ContentCard now clips its children correctly

## 0.2.1+1

- Removed unneeded dependencies

## 0.2.1

- BREAKING: Switch -> EufemiaSwitch
- BREAKING: Slider -> EufemiaSlider

## 0.2.0+3

- Tweaks to BottomTabBar padding

## 0.2.0+2

- Added option to un-center leading widgets in a Cell
- Expanded touch area on bottom app bar items

## 0.2.0+1

- Tweaked BottomAppBarItem padding

## 0.2.0

- New colors, font and tweaks

## 0.1.3+4

- Added contentPadding parameter to Cell

## 0.1.3+3

- Added shrink parameter to Switches

## 0.1.3+2

- Added bottom padding parameter to all lists

## 0.1.3+1

- Removed duplicate bottom border in StaticList

## 0.1.3

- Added static lists

## 0.1.2+6

- Fixed a bug causing Hero widgets to be duplicated in lists

## 0.1.2+5

- Changed the behavior of ScrollableList and StaticList with Cells and children to automatically add a border to the end of the list

## 0.1.2+4

- Exposing height of PaymentCard

## 0.1.2+3

- Added padding to trailing text in cells

## 0.1.2+2

- Added parameters for content card borders and separators

## 0.1.2+1

- Added shadow parameter for content cards

## 0.1.2

- Added company card designs

## 0.1.1+2

- Fixed alignment of trailing widgets combined with implied navigation

## 0.1.1+1

- Ability to combine trailing widget and implied navigation in Cells

## 0.1.1

- Added a new font
- Added assets/splash screens to example app
- Added some system/native controls
- Dark mode optimalizations

## 0.1.0+1

- Fixed the Kenyan flag

## 0.1.0

- Initial release
- Most components/graphics/icons/typography implemented
- Theming & dark mode is still WIP


lib/src/theme.dart
import 'package:eufemia/eufemia.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

const double _iconSize = 16.0;
const double _dialogBorderRadius = 8.0;
const String dnbFont = 'DNB Sans';
const String dnbMonoFont = 'DNB Mono';
final Color _statusbarDarkColor = Colors.transparent;
final Color _statusbarLightColor = Colors.transparent;
final Color _appBarLightColor = Colors.white;
final Color _appBarLightIconColor = EufemiaColors.seaGreenAlt;
final Color _darkIconColor = EufemiaColors.mintGreen;

class Eufemia {
  static ThemeData _lightThemeShared() {
    return ThemeData(
      brightness: Brightness.light,
      primaryColor: EufemiaColors.mintGreen,
      accentColor: EufemiaColors.emeraldGreen,
      appBarTheme: AppBarTheme(
        elevation: 0,
        brightness: Brightness.light,
        color: _appBarLightColor,
        iconTheme: IconThemeData(
          color: _appBarLightIconColor,
          size: _iconSize,
        ),
        actionsIconTheme: IconThemeData(
          color: _appBarLightIconColor,
          size: _iconSize,
        ),
      ),
      iconTheme: IconThemeData(
        size: _iconSize,
        color: EufemiaColors.seaGreenAlt,
      ),
      fontFamily: dnbFont,
      textTheme: lightTextTheme,
      bottomAppBarColor: Colors.white,
      toggleableActiveColor: EufemiaColors.seaGreenAlt,
      dialogTheme: DialogTheme(
        shape: RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(_dialogBorderRadius),
        ),
      ),
      buttonTheme: ButtonThemeData(
        highlightColor: EufemiaColors.oceanGreen.withOpacity(0.2),
        splashColor: EufemiaColors.oceanGreen.withOpacity(0.1),
      ),
      textSelectionColor: EufemiaColors.mintGreenAlt,
      textSelectionHandleColor: EufemiaColors.emeraldGreen,
      cursorColor: EufemiaColors.seaGreenAlt,
      buttonColor: EufemiaColors.seaGreenAlt,
      snackBarTheme: SnackBarThemeData(
        actionTextColor: EufemiaColors.mintGreen,
        backgroundColor: EufemiaColors.coal,
        contentTextStyle: Eufemia.subhead.copyWith(
          color: Colors.white,
        ),
        elevation: 0.1,
        behavior: SnackBarBehavior.floating,
      ),
      inputDecorationTheme: InputDecorationTheme(
        enabledBorder: OutlineInputBorder(
          borderRadius: BorderRadius.circular(4.0),
          borderSide: BorderSide(
            width: 0.5,
            color: EufemiaColors.darkGray,
          ),
        ),
        focusedBorder: OutlineInputBorder(
          borderRadius: BorderRadius.circular(4.0),
          borderSide: BorderSide(
            color: EufemiaColors.seaGreenAlt,
            width: 1.5,
          ),
        ),
        errorBorder: OutlineInputBorder(
          borderRadius: BorderRadius.circular(4.0),
          borderSide: BorderSide(
            color: EufemiaColors.fireRed,
            width: 0.5,
          ),
        ),
        focusedErrorBorder: OutlineInputBorder(
          borderRadius: BorderRadius.circular(4.0),
          borderSide: BorderSide(
            color: EufemiaColors.fireRed,
            width: 1.5,
          ),
        ),
      ),
      primaryTextTheme: lightTextTheme,
    );
  }

  static ThemeData _darkThemeShared() {
    return ThemeData(
      brightness: Brightness.dark,
      primaryColor: EufemiaColors.emeraldGreen,
      accentColor: EufemiaColors.mintGreen,
      canvasColor: Colors.black,
      cardColor: EufemiaColors.gray5,
      iconTheme: IconThemeData(
        size: _iconSize,
        color: EufemiaColors.mintGreen,
      ),
      appBarTheme: AppBarTheme(
        elevation: 0,
        brightness: Brightness.dark,
        color: EufemiaColors.gray6,
        textTheme: darkTextTheme.apply(
          displayColor: EufemiaColors.mintGreen,
          bodyColor: EufemiaColors.mintGreen,
        ),
        iconTheme: IconThemeData(color: _darkIconColor, size: _iconSize),
        actionsIconTheme: IconThemeData(color: _darkIconColor, size: _iconSize),
      ),
      snackBarTheme: SnackBarThemeData(
        actionTextColor: EufemiaColors.mintGreen,
        backgroundColor: EufemiaColors.coal,
        contentTextStyle: Eufemia.subhead.copyWith(
          color: Colors.white,
        ),
        elevation: 0.1,
        behavior: SnackBarBehavior.floating,
      ),
      textTheme: darkTextTheme,
      bottomAppBarColor: Colors.black,
      toggleableActiveColor: EufemiaColors.mintGreen,
      dialogTheme: DialogTheme(
        shape: RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(_dialogBorderRadius),
        ),
      ),
      toggleButtonsTheme: ToggleButtonsThemeData(
        selectedColor: EufemiaColors.mintGreen,
      ),
      textSelectionColor: EufemiaColors.mintGreenAlt.withOpacity(0.5),
      textSelectionHandleColor: EufemiaColors.mintGreen,
      buttonTheme: ButtonThemeData(
        highlightColor: EufemiaColors.mintGreenAlt.withOpacity(0.2),
        splashColor: EufemiaColors.mintGreenAlt.withOpacity(0.1),
      ),
      buttonColor: EufemiaColors.mintGreen,
      inputDecorationTheme: InputDecorationTheme(
        enabledBorder: OutlineInputBorder(
          borderRadius: BorderRadius.circular(4.0),
          borderSide: BorderSide(
            width: 0.5,
            color: EufemiaColors.darkGray,
          ),
        ),
        focusedBorder: OutlineInputBorder(
          borderRadius: BorderRadius.circular(4.0),
          borderSide: BorderSide(
            color: EufemiaColors.mintGreen,
            width: 1.5,
          ),
        ),
        errorBorder: OutlineInputBorder(
          borderRadius: BorderRadius.circular(4.0),
          borderSide: BorderSide(
            color: EufemiaColors.fireRed,
            width: 0.5,
          ),
        ),
        focusedErrorBorder: OutlineInputBorder(
          borderRadius: BorderRadius.circular(4.0),
          borderSide: BorderSide(
            color: EufemiaColors.fireRed,
            width: 1.5,
          ),
        ),
      ),
      primaryTextTheme: darkTextTheme,
    );
  }

  static TextTheme get lightTextTheme {
    return TextTheme(
      headline4: titleLarge.copyWith(color: EufemiaColors.coal),
      headline5: titleMedium.copyWith(color: EufemiaColors.coal),
      headline6: bodyEmphasized.copyWith(color: EufemiaColors.coal),
      subtitle1: subhead.copyWith(color: EufemiaColors.coal),
      bodyText2: body.copyWith(color: EufemiaColors.coal),
      bodyText1: bodyEmphasized.copyWith(color: EufemiaColors.coal),
      caption: caption.copyWith(color: EufemiaColors.coal),
      subtitle2: footnote.copyWith(color: EufemiaColors.darkGray),
    );
  }

  static TextTheme get darkTextTheme {
    return TextTheme(
      headline4: titleLarge.copyWith(color: Colors.white),
      headline5: titleMedium.copyWith(color: Colors.white),
      headline6: bodyEmphasized.copyWith(color: Colors.white),
      subtitle1: subhead.copyWith(color: Colors.white),
      bodyText2: body.copyWith(color: Colors.white),
      bodyText1: bodyEmphasized.copyWith(color: Colors.white),
      caption: caption.copyWith(color: Colors.white),
      subtitle2: footnote.copyWith(color: Colors.white.withOpacity(0.75)),
    );
  }

  // Text styles

  ///Large button
  static TextStyle get buttonLarge => TextStyle(
        fontSize: 16.0,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  static TextStyle get buttonSmall => TextStyle(
        fontSize: 14.0,
        fontWeight: FontWeight.w500,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  /// 34pt Bold
  static TextStyle get titleLarge => TextStyle(
        fontWeight: FontWeight.bold,
        fontSize: 34,
        height: 1.29,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  /// 24pt Regular
  static TextStyle get titleMedium => TextStyle(
        fontSize: 24,
        height: 1.33,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  /// 24pt Bold
  static TextStyle get titleMediumEmphasized => TextStyle(
        fontWeight: FontWeight.bold,
        fontSize: 24,
        height: 1.33,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  /// 20pt Regular
  static TextStyle get title => TextStyle(
        fontSize: 20,
        height: 1.4,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  /// 20pt Semibold
  static TextStyle get titleDemi => TextStyle(
        fontWeight: FontWeight.w500,
        fontSize: 20,
        height: 1.4,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  /// 20pt Bold
  static TextStyle get titleEmphasized => TextStyle(
        fontWeight: FontWeight.bold,
        fontSize: 20,
        height: 1.4,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  ///  17pt Regular
  static TextStyle get body => TextStyle(
        fontSize: 17,
        height: 1.41,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  /// 17pt Bold
  static TextStyle get bodyEmphasized => TextStyle(
        fontWeight: FontWeight.bold,
        fontSize: 17,
        height: 1.41,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  /// 15pt Regular
  static TextStyle get subhead => TextStyle(
        fontSize: 15,
        height: 1.33,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  /// 15pt Bold
  static TextStyle get subheadEmphasized => TextStyle(
        fontWeight: FontWeight.bold,
        fontSize: 15,
        height: 1.33,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  /// 13pt Regular
  static TextStyle get footnote => TextStyle(
        fontSize: 13,
        height: 1.23,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  /// 13pt Bold
  static TextStyle get footnoteEmphasized => TextStyle(
        fontWeight: FontWeight.bold,
        fontSize: 13,
        height: 1.23,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  /// 11pt Regular
  static TextStyle get caption => TextStyle(
        fontSize: 11,
        height: 1.45,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  /// 11pt Bold
  static TextStyle get captionEmphasized => TextStyle(
        fontWeight: FontWeight.bold,
        fontSize: 11,
        height: 1.45,
        package: 'eufemia',
        fontFamily: 'DNB Sans',
      );

  /// Returns the default Eufemia light theme, and sets the status bar brightness.
  ///
  /// [statusBarColor]: Sets the status bar color. Defaults to [Colors.transparent]
  static ThemeData lightTheme({Color statusBarColor}) {
    setLightStatusBar(statusBarColor: statusBarColor);
    return _lightThemeShared();
  }

  /// Returns the default Eufemia dark theme, and sets the status bar brightness.
  ///
  /// [statusBarColor]: Sets the status bar color. Defaults to [Colors.transparent]
  static ThemeData darkTheme({Color statusBarColor}) {
    setDarkStatusBar(statusBarColor: statusBarColor);
    return _darkThemeShared();
  }

  /// Sets the status bar brightness.
  ///
  /// [statusBarColor]: Sets the status bar color. Defaults to [Colors.transparent]
  static void setLightStatusBar({Color statusBarColor}) {
    statusBarColor ??= _statusbarLightColor;
    SystemChrome.setSystemUIOverlayStyle(getOverlayStyle(
      Brightness.light,
      statusBarColor,
    ));
  }

  /// Sets the status bar brightness.
  ///
  /// [statusBarColor]: Sets the status bar color. Defaults to [Colors.transparent]
  static void setDarkStatusBar({Color statusBarColor}) {
    statusBarColor ??= _statusbarDarkColor;
    SystemChrome.setSystemUIOverlayStyle(getOverlayStyle(
      Brightness.dark,
      statusBarColor,
    ));
  }

  /// Returns a [SystemUiOverlayStyle] object which can be used with [SystemChrome] to
  /// well, set the System UI overlay style.
  ///
  /// [brightness]: The brightness of the status bar. Affects the icon brightness accordingly.
  /// [statusBarColor]: The color of the status bar. Should reflect the brightness given, e.g.
  /// a dark color for [Brightness.dark] to make the icons visible.
  static SystemUiOverlayStyle getOverlayStyle(
    Brightness brightness,
    Color statusBarColor,
  ) {
    if (brightness == Brightness.dark) {
      return SystemUiOverlayStyle(
        // Only affects Android
        statusBarColor: statusBarColor,
        statusBarBrightness: Brightness.dark,
        statusBarIconBrightness: Brightness.dark,
        systemNavigationBarIconBrightness: Brightness.dark,
      );
    } else {
      return SystemUiOverlayStyle(
        // Only affects Android
        statusBarColor: statusBarColor,
        statusBarBrightness: Brightness.light,
        statusBarIconBrightness: Brightness.light,
        systemNavigationBarIconBrightness: Brightness.light,
      );
    }
  }

  static InputBorder unfocusedWithTextBorder(BuildContext context) {
    return context.theme.inputDecorationTheme.focusedBorder.copyWith(
      borderSide: BorderSide(
        width: 0.5,
        color:
            context.theme.inputDecorationTheme.focusedBorder.borderSide.color,
      ),
    );
  }
}
